packages/stark-core/src/common/bootstrap/abstract-stark-main.ts

```diff
@@ -0,0 +1,157 @@
+import {ApplicationRef, ComponentRef, enableProdMode, NgModuleRef} from "@angular/core";
+import {disableDebugTools, enableDebugTools} from "@angular/platform-browser";
+import {createNewHosts} from "@angularclass/hmr";
+
+import {StarkMain} from "./stark-main.intf";
+import {StarkEnvironment} from "environments/environment";
+
+/**
+ * Parent class for bootstrapping Stark applications.
+ * It handles the DOMContentLoaded event of the Web browser in order to bootstrap the application.
+ * This class also supports HMR in development mode.
+ */
+export abstract class AbstractStarkMain implements StarkMain {
+	protected constructor(protected environment: StarkEnvironment) {
+		// no-op
+	}
+
+	/**
+	 * Must be implemented by children classes. Called within bootstrapDomReady (i.e., when the DOM has been loaded)
+	 */
+	public abstract main: () => Promise<any>;
+
+	public mainWrapper = (): void => { // written like this because otherwise "this" will not be captured :)
+		// no need for the event listener after this
+		document.removeEventListener("DOMContentLoaded", this.mainWrapper, false);
+
+		if (this.environment.hmr) {
+			if (module["hot"]) {
+				this.bootstrapHmr(module, this.main);
+			} else {
+				console.error("HMR is not enabled for webpack-dev-server!");
+			}
+		} else {
+			this.invokeMain();
+		}
+	};
+	
+	private invokeMain():void {
+		// FIXME handle errors nicely (UX!)
+		this.main().catch((err: any) => console.error(err));
+
+		// TODO reintroduce later
+		// bootstrap the app as long as the app root element is there (it won't be there in case the browser is outdated)
+		// if (document.getElementsByClassName("stark-app")[0]) {
+		// 	// add anything relevant here
+		// } else {
+		// 	console.error("Could not bootstrap the App: unsupported browser.");
+		// }
+	}
+
+	/**
+	 * Calls the main function once the DOM has loaded
+	 */
+	protected bootstrapDomReady(): void {
+		switch (document.readyState) {
+			case "loading":
+				document.addEventListener("DOMContentLoaded", this.mainWrapper, false);
+				break;
+			case "interactive":
+			case "complete":
+			default:
+				if (this.environment.hmr) {
```

Comment:
Perhaps we can extract this entire block to a function because it is also used in the `mainWrapper` property, what do you think?

Reply:
Indeed